File: tests/astro/python/test_ephemeris.py
#!/usr/bin/env python3

import datetime
import ephem

year   = 2000
month  = 1
day    = 1
hour   = 0
minute = 0
second = 0
dt = datetime.datetime(year, month ,day, hour, minute, second)
sun = ephem.Sun()
moon = ephem.Moon()
sun.compute(dt, epoch='2000')
moon.compute(dt, epoch='2000')

print()
print("date: ", str(ephem.Date(dt)))
print("jd: ", str(ephem.julian_date(dt)))

print("Sun:")
print("RA: ", float(sun.ra))
print("dec: ", float(sun.dec))

print("Moon:")
print("RA: ", float(moon.ra))
print("dec: ", float(moon.dec))

year   = 2024
month  = 4
day    = 24
hour   = 17
minute = 15
second = 30
dt = datetime.datetime(year, month ,day, hour, minute, second)
sun = ephem.Sun()
moon = ephem.Moon()
sun.compute(dt, epoch='2000')
moon.compute(dt, epoch='2000')

print()
print("date: ", str(ephem.Date(dt)))
print("jd: ", str(ephem.julian_date(dt)))

print("Sun:")
print("RA: ", float(sun.ra))
print("dec: ", float(sun.dec))

print("Moon:")
print("RA: ", float(moon.ra))
print("dec: ", float(moon.dec))

File: tests/astro/TestEphemeris.cpp
#include <gtest/gtest.h>

#include <mcutils/astro/Ephemeris.h>
#include <mcutils/misc/Units.h>

class TestEphemeris : public ::testing::Test
{
protected:
    TestEphemeris() {}
    virtual ~TestEphemeris() {}
    void SetUp() override {}
    void TearDown() override {}
};

TEST_F(TestEphemeris, CanComputeSun)
{
    // expected values calculated with PyEphem
    // tests/astro/python/test_ephemeris.py

    mc::DateTime dt;
    mc::Ephemeris ephem;

    double tol_rad = mc::Units::deg2rad(0.01);

    dt.year   = 2000;
    dt.month  = 1;
    dt.day    = 1;
    dt.hour   = 0;
    dt.minute = 0;
    dt.second = 0;
    dt.msec   = 0;
    ephem.Update(dt, 0.0, 0.0);
    EXPECT_NEAR(ephem.sun().ra_dec.ra(), 4.899595716773554, tol_rad);
    EXPECT_NEAR(ephem.sun().ra_dec.dec(), -0.40266683881073745, tol_rad);

    dt.year   = 2024;
    dt.month  = 4;
    dt.day    = 24;
    dt.hour   = 17;
    dt.minute = 15;
    dt.second = 30;
    dt.msec   = 0;
    ephem.Update(dt, 0.0, 0.0);
    EXPECT_NEAR(ephem.sun().ra_dec.ra(), 0.571160439394054, tol_rad);
    EXPECT_NEAR(ephem.sun().ra_dec.dec(), 0.23021727848778714, tol_rad);
}

TEST_F(TestEphemeris, CanComputeMoon)
{
    // expected values calculated with PyEphem
    // tests/astro/python/test_ephemeris.py

    mc::DateTime dt;
    mc::Ephemeris ephem;

    double tol_rad = mc::Units::deg2rad(0.1);

    dt.year   = 2000;
    dt.month  = 1;
    dt.day    = 1;
    dt.hour   = 0;
    dt.minute = 0;
    dt.second = 0;
    dt.msec   = 0;
    ephem.Update(dt, 0.0, 0.0);
    EXPECT_NEAR(ephem.moon().ra_dec.ra(), 3.7815961040162467, tol_rad);
    EXPECT_NEAR(ephem.moon().ra_dec.dec(), -0.1569823363895957, tol_rad);

    dt.year   = 2024;
    dt.month  = 4;
    dt.day    = 24;
    dt.hour   = 17;
    dt.minute = 15;
    dt.second = 30;
    dt.msec   = 0;
    ephem.Update(dt, 0.0, 0.0);
    EXPECT_NEAR(ephem.moon().ra_dec.ra(), 3.839393709858737, tol_rad);
    EXPECT_NEAR(ephem.moon().ra_dec.dec(), -0.31637240994389354, tol_rad);
}
